### DISCUSSION TITLE
Way to counteract condensed table view?

### DISCUSSION THREAD

**Discussion by Alex-Wenner-FHR** [h2oai/wave#851] [Q&A] [answered]
![image](https://user-images.githubusercontent.com/78372056/120002400-68db3f80-bf9a-11eb-9cda-210c4f87b43f.png)

I tried adjusting the size of the table but that doesn't seem to fix an issue like this when user filters and we see a singular record. Gets very small and hard to see! Any way to fix this currently?

**Comment by mturoci** [comment, marked as answer]
Hi @AlexWennerDT, could you please add a minimal repro? Our example seems fine:
![image](https://user-images.githubusercontent.com/64769322/122229903-6e50e900-ceb9-11eb-9c33-a9dadcfbd0b7.png)

**Reply by Alex-Wenner-FHR** [reply]
Sure! I will try to put together a small snippet of my code using dummy data. the main difference I see here is maybe the progress feature and the fact that your table is 1/100 whereas mine is 1/1 due to manual filtering.

**Reply by Alex-Wenner-FHR** [reply]
Here you are! @mturoci 

The singular record view is a little bit more viewable than my real code, but you can still see the 1 record scroll effect.

```py
from h2o_wave import ui, app, Q, on, handle_on, main
import pandas as pd
from sklearn.datasets import make_blobs




X, y = make_blobs(n_samples=100, centers=2, n_features=4, random_state=0)
dataframe = pd.concat([pd.DataFrame(X), pd.DataFrame(y)], axis=1)
dataframe.columns = ['A', 'B', 'C', 'D', 'E']
dataframe = dataframe.round(decimals=5)
dataframe = dataframe.astype(str)
df_renamed = dataframe



@app('/')
async def serve(q: Q):
    print(df_renamed.columns)
    if not q.client.initalized:
        init_client(q)
    else:
        await handle_on(q)
        
    await q.page.save()
    print('served')
    
def init_client(q: Q):
    if not q.app.initalized:
        init_app(q)
    render_table(q, df_renamed)
    q.client.initalized = True


def init_app(q: Q):
    q.app.initalized = True


@on('filter_B')
async def filter_0(q: Q):
    await filter_data(q, 'B')

@on('filter_C')
async def filter_1(q: Q):
    await filter_data(q, 'C')

@on('filter_D')
async def filter_2(q: Q):
    await filter_data(q, 'D')


async def filter_data(q: Q, key:str):
    q.client[key] = q.args[f'filter_{key}']
    print(q.client[key])
    print(df_renamed[key])
    df = df_renamed[df_renamed[key].isin(q.client[key])]
    print(df)
    await update_table(q, df)

async def update_table(q, df):
    q.page['table'].items[0].table.rows = [
        ui.table_row(
            str(i),
            [str(row[name]) for name in df.columns]
        ) for i, row in df.iterrows()
    ]

def render_filter(q: Q, filter_name):
    print('inside render_filter')
    picker = [ui.picker(
            name=f'filter_{filter_name}',
            label=f'{filter_name}',
            choices=[ui.choice(
                    name=str(d),
                    label=str(d)
                    )for d in df_renamed[filter_name].unique()],
            max_choices=5,
            values=q.args[filter_name],
            trigger=True,
        ) for filter_name in ['B', 'C', 'D'] 
    ]

    q.page['filters'] = ui.form_card(
        box ='1 2 3 7',
        items = [*picker]
        )
    
def render_table(q: Q, df):

    for f in ['B', 'C', 'D']:
        render_filter(q, f)
    print('rendering table')
    q.page['table'] = ui.form_card(
        box='4 2 8 7',
        items=[
            ui.table(name='table',
                     downloadable=True,
                     resettable = True,
                     columns=[ui.table_column(
                         name=c,
                         label=c,
                         filterable = True,
                         searchable= True,
                     ) for c in df.columns],
                     rows=[ui.table_row(
                         str(i),
                         [str(row[name]) for name in df.columns]
                     ) for i, row in df.iterrows()],
                     groupable=True,
                     values=[''])
        ]
    )

    print('table rendered')



```

**Reply by mturoci** [reply]
Thanks for the repro @AlexWennerDT! Very helpful. Can you try setting height of the table like this?
```py
ui.table(
    name='table',
    downloadable=True,
    resettable=True,
    columns=[ui.table_column(name=c, label=c, filterable=True, searchable=True) for c in df.columns],
    rows=[ui.table_row(str(i), [str(row[name]) for name in df.columns]) for i, row in df.iterrows()],
    groupable=True,
    height='580px', # Set fixed table height.
    values=['']
)
```

**Reply by Alex-Wenner-FHR** [reply]
Thanks for the tip @mturoci, that solved the issue!

**Reply by azim-b** [reply]
@mturoci I'm seeing a similar problem. Is it possible to set the table height to be dynamic so it changes with the browser size, and stops just before the page length is complete so that the footer card is visible without scrolling?


Here's what the table looks like:

![image](https://user-images.githubusercontent.com/64787868/137778034-85cdd572-cd05-46cb-937f-05382617051b.png)


And these are the code snippets for reference:

![image](https://user-images.githubusercontent.com/64787868/137778346-b364b417-f828-4e62-92a8-e490b5631a58.png)

![image](https://user-images.githubusercontent.com/64787868/137778458-9d927323-baec-43c0-8e0a-b8f3fcee85d0.png)

**Reply by Alex-Wenner-FHR** [reply]
@azim-b I actually ran into the same issue and ended up using vh/vw … not sure if this is best practice but it seemed to work for me!

**Reply by mturoci** [reply]
@azim-b @AlexWennerDT see https://github.com/h2oai/wave/discussions/1074 for general solution.

However, `ui.table` is a bit special and in order to achieve a truly pixel perfect solution, you would need to use CSS `calc` function.

The solution for now would be to specify table height like `height='calc(100vh - headerHeightInPx - navigateheightInPx - footerHeightInPx)', where `100vh` means 100% of viewport height and rest of the values are obvious.

We might make this work out of the box one day (using the technique linked above), but need to solve some tech problems first - especially plots.
